### PR TITLE
Base method module satellite

### PR DESCRIPTION
**Pull request by AkaMrmuffin**
1. Updated the satellite module according to the latest version of LDAR Sim. 
2. Created the orbit base method.
3. Created a satellite sensor.

### CODE REVIEW

File: LDAR_Sim/model_code/dev_satellite_company.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+from methods.company import BaseCompany
+from methods.crew import BaseCrew
+
+
+class dev_satellite_company(BaseCompany):
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/dev_satellite_company.py:5]: You can actually forgo using dev_satellite_company.py and just call dummy_company (P_dev_aircraft). In the future once we have all of the methods updated, we will remove dummy company.

File: LDAR_Sim/model_code/generic_functions.py [hunk 2]
```diff
@@ -291,3 +295,70 @@ def quick_cal_daylight(date, lat, lon):
     sunset = ss
 
     return (sunrise, sunset)
+
+def ecef_to_llh(ecef_km):
+    """
+    covert ecef cooridnate system to latitude, longitude and altitude
+    """
+    a = 6378.1370
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/generic_functions.py:303]: Either use descriptive variable name or a comment; What are these for?

File: LDAR_Sim/model_code/generic_functions.py [hunk 3]
```diff
@@ -291,3 +295,70 @@ def quick_cal_daylight(date, lat, lon):
...
+        pt4 = (lon2-0.1, lat2)
+
+        # create that coverage area polygon
+        polygon3 = Polygon([pt1, pt2, pt3, pt4])
+        Poly.append(polygon3)
+        Day.append(T1)
+
+        T1 += datetime.timedelta(minutes=interval)
+
+    return Day, Poly
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/generic_functions.py:364]: should be lower case

File: LDAR_Sim/model_code/methods/deployment/orbit_company.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
+from methods.deployment._base import SchedCompany as BaseSchedCompany
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/methods/deployment/orbit_company.py:1]: Add docstrings, and copyright licensing where possible.

File: LDAR_Sim/model_code/methods/deployment/orbit_company.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    def __init__(self, config, parameters, state):
+        self.parameters = parameters
+        self.config = config
+        self.state = state
+
+        # --- inherited ---
+        # base.company ->  get_deployment_dates()
+        # base.company ->  can_deploy_today()
+
+    def get_deployment_dates(self):
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/methods/deployment/orbit_company.py:14]: This is already part of the base .company :
`  --- inherited ---`
 `base.company ->  get_deployment_dates()`
` base.company ->  can_deploy_today()`

File: LDAR_Sim/model_code/methods/deployment/orbit_company.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        else:
+            self.deployment_years = list(
+                range(self.state['t'].start_date.year, self.state['t'].end_date.year+1))
+
+        if len(self.config['scheduling']['deployment_months']) > 0:
+            self.deployment_months = self.config['scheduling']['deployment_months']
+        else:
+            self.deployment_months = list(range(1, 13))
+
+    def can_deploy_today(self, date):
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/methods/deployment/orbit_company.py:31]: same as above

File: LDAR_Sim/model_code/methods/deployment/orbit_crew.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
+import numpy as np
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/methods/deployment/orbit_crew.py:1]: Add docstrings, copyright and licensing where applicable.

File: LDAR_Sim/model_code/methods/deployment/orbit_crew.py [hunk 8]
```diff
@@ -30,7 +30,7 @@ def __init__(self, id, lat, lon, state, config, parameters, deployment_days, hom
         # obtain TLE file path
         m_name = self.config['label']
         self.sat = self.parameters['methods'][m_name]['satellite_name']
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/methods/deployment/orbit_crew.py:32]: self.parameters['methods'][m_name] can be replaced with self.config, they are the same thing. That's done elsewhere in the code occasionally but should be made consistent.

File: LDAR_Sim/model_code/methods/deployment/orbit_crew.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+            if x == self.sat:
+                break
+            i += 1
+        TLE_LINES = (TLEs[i+1], TLEs[i+2])
+        self.predictor = get_predictor_from_tle_lines(TLE_LINES)
+        return
+
+    def get_orbit_path(self):
+
+        #### initiate the orbit path ####
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/methods/deployment/orbit_crew.py:152]: multiple consecutive # is a pep8 error. I've been replacing with # --- Text  --- for heading comments,  but also not very consistently.

File: LDAR_Sim/model_code/methods/sensors/Satellite_sensor.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    m_name = self.config['label']
+    # extract the wind speed on site based on site's geo indices
+    site_lat = np.float16(site['lat'])
+    site_lon = np.float16(site['lon'])
+    lat_idx = geo_idx(site_lat, self.state['weather'].latitude)
+    lon_idx = geo_idx(site_lon, self.state['weather'].longitude)
+    ti = self.state['t'].current_timestep
+    windspeed = self.state['weather'].winds[ti, lat_idx, lon_idx]
+    # Calculate MDL
+    Q_min = 5.79 * (1.39/windspeed)
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/methods/sensors/Satellite_sensor.py:16]: converting into scf/km/hr? what are the conversion factors.

File: LDAR_Sim/model_code/satellite.py [hunk 11]
```diff
@@ -156,39 +158,47 @@ def assess_weather(self, site):
...
+        site_lat = np.float16(site['lat'])
+        site_lon = np.float16(site['lon'])
+        lat_idx = geo_idx(site_lat, self.state['weather'].latitude)
+        lon_idx = geo_idx(site_lon, self.state['weather'].longitude)
+        windspeed = self.state['weather'].winds
+
+        ti = self.state['t'].current_timestep
+        U = windspeed[ti, lat_idx, lon_idx]
+        Q_min = 5.79 * (1.39/U)
+        # set MDL to 0 for testing #####
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/satellite.py:194]: Multiple # is a pep8 error.

File: LDAR_Sim/model_code/satellite.py [hunk 12]
```diff
@@ -217,16 +229,14 @@ def choose_site(self):
                 # If the site hasn't been attempted yet today
                 if not site['attempted_today_satellite?']:
 
-                    # If the site is 'unripened' (i.e. hasn't met the minimum interval set out in
-                    # the LDAR regulations/policy), break out - no LDAR today
-                    if site['satellite_t_since_last_LDAR'] \
-                            < self.parameters['methods']['satellite']['min_interval']:
-                        self.state['t'].current_date = self.state['t'].current_date.replace(hour=23)
+                    # If the site is 'unripened' (i.e. hasn't met the minimum interval set out in the LDAR regulations/policy), break out - no LDAR today
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/satellite.py:232]: line is too long

File: LDAR_Sim/model_code/satellite.py [hunk 13]
```diff
@@ -217,16 +229,14 @@ def choose_site(self):
                 # If the site hasn't been attempted yet today
                 if not site['attempted_today_satellite?']:
 
-                    # If the site is 'unripened' (i.e. hasn't met the minimum interval set out in
-                    # the LDAR regulations/policy), break out - no LDAR today
-                    if site['satellite_t_since_last_LDAR'] \
-                            < self.parameters['methods']['satellite']['min_interval']:
-                        self.state['t'].current_date = self.state['t'].current_date.replace(hour=23)
+                    # If the site is 'unripened' (i.e. hasn't met the minimum interval set out in the LDAR regulations/policy), break out - no LDAR today
+                    if site['satellite_t_since_last_LDAR'] < self.parameters['methods']['satellite']['min_interval']:
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/satellite.py:233]: line is too long

File: LDAR_Sim/model_code/satellite.py [hunk 14]
```diff
@@ -261,13 +271,13 @@ def visit_site(self, facility_ID, site):
 
         # Add vented emissions
         venting = 0
-        if self.parameters['consider_venting']:
+        if self.parameters['consider_venting'] == True:
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/satellite.py:274]: pep8 error, use if self.parameters['consider_venting']:

File: LDAR_Sim/model_code/satellite_company.py [hunk 15]
```diff
@@ -110,7 +92,7 @@ def __init__(self, state, parameters, config, timeseries, module_name):
 
     def deploy_crews(self):
         """
-        The satellite company tells all the crews to get to work.
+        The aircraft company tells all the crews to get to work.
```
Comment (KeeganShaw-GIS) [inline LDAR_Sim/model_code/satellite_company.py:95]: satellite

**Review by KeeganShaw-GIS — CHANGES_REQUESTED** [review]
@AkaMrmuffin Nice work Mo, just a few changes required. Can be merged shortly after. Coleman and I finished making the Stationary branch. You could steal some of the docstrings and copyright licensing from there if you want as we did a bit of updating on the template / mobile documentation.

File: LDAR_Sim/model_code/methods/deployment/orbit_crew.py (hunk 8 above)
Comment (AkaMrmuffin) [reply LDAR_Sim/model_code/methods/deployment/orbit_crew.py:32]: fixed

File: LDAR_Sim/model_code/dev_satellite_company.py (hunk 1 above)
Comment (AkaMrmuffin) [reply LDAR_Sim/model_code/dev_satellite_company.py:5]: fixed

File: LDAR_Sim/model_code/generic_functions.py (hunk 3 above)
Comment (AkaMrmuffin) [reply LDAR_Sim/model_code/generic_functions.py:364]: Fixed

File: LDAR_Sim/model_code/generic_functions.py (hunk 2 above)
Comment (AkaMrmuffin) [reply LDAR_Sim/model_code/generic_functions.py:309]: fixed

File: LDAR_Sim/model_code/satellite.py (hunk 12 above)
Comment (AkaMrmuffin) [reply LDAR_Sim/model_code/satellite.py:232]: Fixed

File: LDAR_Sim/model_code/satellite.py (hunk 13 above)
Comment (AkaMrmuffin) [reply LDAR_Sim/model_code/satellite.py:233]: Fixed

File: LDAR_Sim/model_code/satellite.py (hunk 11 above)
Comment (AkaMrmuffin) [reply LDAR_Sim/model_code/satellite.py]: Fixed

File: LDAR_Sim/model_code/satellite.py (hunk 14 above)
Comment (AkaMrmuffin) [reply LDAR_Sim/model_code/satellite.py]: Fixed

**Pull request merged**